### ISSUE TITLE
Rule proposal: `class-method-style`

### ISSUE THREAD

**Issue by zanminkian** [sindresorhus/eslint-plugin-unicorn#2504]
### Description

We have 2 styles to define a method in class:

```js
class Foo {
  methodStyle() {
    console.log("I am method style method");
  }
  propertyStyle1 = () => {
    console.log("I am property style method");
  };
  propertyStyle2 = function () {
    console.log("I am property style method");
  };
}
```

I recommend method style and disallow property style. Therefore, this rule `class-method-style` accept an option to specify the style, `method` (default) or `property`.

### Fail

```js
class Foo {
  propertyStyle1 = () => {
    console.log("I am property style method");
  };
}
```

```js
class Foo {
  propertyStyle2 = function () {
    console.log("I am property style method");
  };
}
```

### Pass

```js
class Foo {
  methodStyle() {
    console.log("I am method style method");
  }
}
```

### Proposed rule name

class-method-style

### Additional Info

Related rule: https://typescript-eslint.io/rules/method-signature-style.
Related issue: #1737

**Comment by fregante** [comment]
Makes sense, as long as the method doesn't use `this`. The only problem that comes to mind is that:

1. User intentionally writes `prop = () => this.doSomething()` ✅
2. User changes it to `prop = () => doSomething()`, which triggers the rule ❌
3. Rule or user changes it to `prop() {return doSomething()}` ✅
4. User later uses `this` again `prop() {return doSomething(this)}` ✅⚠️

In the 4th step, everything looks fine, but the original intent of the method is lost.

I wonder the the rule should go both ways and:

- force `prop() {}` when it doesn't detect `this`
- force `prop: () => {}` when it detects `this`

**Comment by fregante** [comment]
I think you're missing the point of the property style: they are pre-bound. 

```js
class A {
  value = 1;
  get = () => console.log(this.value)
}

const a = new A()
setTimeout(a.get, 1000)
```

This will log `1`. If you use a method, it will log `undefined`

There are only a few practical reasons to prefer methods over properties:

- methods are faster to instantiate (if I remember correctly)
- methods can be bound to whatever the caller prefers

What you're describing are just stylistic preferences, but this has actual runtime differences.

**Comment by sindresorhus** [comment]
I dont think we should add this.

This is not just a style choice. Class field functions and class methods have different runtime behavior: own property vs prototype method, lexical vs dynamic `this`, per-instance identity, enumerability, initialization timing, and inheritance behavior.

The similarity to `prefer-short-arrow-method` is exactly why that rule intentionally only applies to object literals, not classes.

For the narrower case of class field functions that don’t use `this`, ESLint’s `class-methods-use-this` already has `enforceForClassFields`.
